File: common/scala/src/main/scala/org/aiddl/common/scala/math/graph/PathExtractor.scala
package org.aiddl.common.scala.math.graph

import scala.collection.mutable.Map
import scala.collection.mutable.HashMap
import scala.collection.immutable.Set
import scala.collection.immutable.HashSet
import org.aiddl.core.scala.function.Function
import org.aiddl.core.scala.function.InterfaceImplementation
import org.aiddl.core.scala.representation._
import org.aiddl.common.scala.Common._

import org.aiddl.core.scala.representation.TermImplicits._
import org.aiddl.core.scala.representation.TermCollectionImplicits.seq2ListTerm

class PathExtractor extends Function with InterfaceImplementation {
    val interfaceUri = Sym("org.aiddl.common.math.graph.path-extractor")

    def apply( args: Term ): Term = args match {
        case Tuple(p, s, g) => { this(s, g, p) } case _ => ???
    }

    def apply( s: Term, g: Term, p: Function ): Term = {
        extract(s, g, p) match {
            case Some(path) if (path.length == 1) => ListTerm.empty
            case Some(path) => ListTerm(path.reverse)
            case None => NIL
        }
    }

    private def extract( s: Term, g: Term, p: Function ): Option[List[Term]] = {
        if ( s == g ) Some(List(s))
        else if (p(g) == NIL) None
        else extract(s, p(g), p) match {
            case Some(r) => Some(g :: r)
            case None => None
        }
    }
}

File: common/java/src/main/java/org/aiddl/common/java/math/graph/PathExtractor.java
package org.aiddl.common.java.math.graph;

import java.util.LinkedList;

import org.aiddl.common.java.CommonTerm;
import org.aiddl.core.java.interfaces.Function;
import org.aiddl.core.java.interfaces.InterfaceImplementation;
import org.aiddl.core.java.representation.SymbolicTerm;
import org.aiddl.core.java.representation.Term;

public class PathExtractor implements Function, InterfaceImplementation {
	
	private final static SymbolicTerm InterfaceUri = Term.sym("org.aiddl.common.math.graph.path-extractor");
	
	@Override
	public SymbolicTerm getInterfaceUri() {
		return InterfaceUri;
	}

	@Override
	public Term apply(Term args) {
		Term predecessors = args.get(0);
		Term start = args.get(1);
		Term goal = args.get(2);

		if ( predecessors.equals(CommonTerm.NIL) ) {
			return CommonTerm.NIL;
		} else if ( start.equals(goal) ) {
			return Term.list();
		} else if ( predecessors.get(goal).equals(CommonTerm.NIL) ) {
			return CommonTerm.NIL;
		}
		
		LinkedList<Term> path = new LinkedList<>();
		Term current = goal;
		while ( !predecessors.get(current).equals(start) ) {
			path.addFirst(current);
			current = predecessors.get(current);
		}
		path.addFirst(current);
		path.addFirst(start);
			
		return Term.list(path);
	}

}
